src/shared/sam/cli/samCliDeploy.ts

```diff
@@ -0,0 +1,46 @@
+/*!
+ * Copyright 2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+'use strict'
+
+import { ChildProcessResult } from '../../utilities/childProcess'
+import { DefaultSamCliProcessInvoker, SamCliProcessInvoker } from './samCliInvoker'
+
+export class SamCliDeployInvocation {
+    public constructor(
```

Comment:
"ECMA6TypeScriptonic" prefer single typed function/constructor arg for safety/clarity/readability.

Reply:
True it's annoying to double type things when defining classes using destructuring params in TypeScript. It appears that inline field declaration has been a crutch for this project that's caused us to write code that is hard to read and understand without looking at the the implementation of every constructor. The project also leans unusually heavily on classes and class hierarchies for a JavaScript project.  In Java it's solid guidance to [favor composition over inheritance](https://en.wikipedia.org/wiki/Composition_over_inheritance) and favor the use of object factories over constructors. This is even more true in JavaScript where object literals are so widely used (and preferred). I don't want to derail this CR, but do want to work towards identifying our standards. Perhaps we should consider favoring interfaces and object literals over classes. In cases where we would find ourselves constructing objects of the same type with the same defaults in several places (one reason we might have reached for classes), we could use object factories. Prefer to extend interfaces not classes. 

Here is an example:

```
interface Person {
  name: string
  shout?(): string
}

interface CEO extends Person {
  companyManaged: string
}

const makePerson = (params: Partial<Person> = {}): Person => {
  const person: Person = {
    name: params.name || 'John Doe',
    shout: params.shout || (() => {
      return `Listen to me, I'm ${person.name}!` 
      // use local variable, person, where 'this' would have be used in a class
    })
  }

  return person
}

const makeCeo = ({
  companyManaged = '',
  ...restParams
}: Partial<CEO>): CEO => {
  return { // Don't need local variable if 'this' wouldn't have been needed
    companyManaged,
    ...makePerson(restParams), // Composition instead of inheritance
  }
}
```

Reply:
In this example, a person, we don't need a factory since developers wouldn't generally want to create multiple people that are the same. This leaves me at the final tenet:

**Consider separating properties from behavior; i.e. using functions rather than instance method.**

One of the common problems associated with OOP is the grab bag of state that's dragged around with every object. When we look at a method, it's not clear which member variables of the object or its parents are used. This is similar to our `env` grab bag. The JS way is to export functions in most cases rather than classes with embedded methods.

In our person example taking these concepts into account, we'd prefer simply:

```
export interface Person {
  name: string
  shout?(): string
}

export interface CEO extends Person {
  companyManaged: string
}

export const shout = (params: {name: string}) => {
  return `Listen to me, I'm ${name}!`
}
```

The results is clear, concise, and trivial to test by declaring functions to take only the precise inputs/outputs used we eliminate the common OOP issue: 

> You wanted a banana but what you got was a gorilla holding the banana and the entire jungle